Avoid using ReduceOp->output_args() in rfactor

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52196 Remove ReduceOp::accumulator
* #52187 [nnc] Remove output_args from ReduceOp
* **#52177 Avoid using ReduceOp->output_args() in rfactor**
* #52170 Dont look at reduction output args when computing mem dependencies

Summary:

I'm trying to get rid of `output_args` for reductions, because they
shouldn't be necessary; it's reducing over its reduction axis, why
does it need to know where its output is going?

Rfactor is probably the trickiest place where we use output_args, but
it looks like it's mostly just carrying around the location of the
store, so use that instead.

Test Plan:
build/bin/test_tensorexpr && build/bin/tensorexpr_bench

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D26420548](https://our.internmc.facebook.com/intern/diff/D26420548)

bertmaher committed:
Avoid using ReduceOp->output_args() in rfactor

Summary:

I'm trying to get rid of `output_args` for reductions, because they
shouldn't be necessary; it's reducing over its reduction axis, why
does it need to know where its output is going?

Rfactor is probably the trickiest place where we use output_args, but
it looks like it's mostly just carrying around the location of the
store, so use that instead.

Test Plan:
build/bin/test_tensorexpr && build/bin/tensorexpr_bench

[ghstack-poisoned]

bertmaher committed:
Update on "Avoid using ReduceOp->output_args() in rfactor"


Summary:

I'm trying to get rid of `output_args` for reductions, because they
shouldn't be necessary; it's reducing over its reduction axis, why
does it need to know where its output is going?

Rfactor is probably the trickiest place where we use output_args, but
it looks like it's mostly just carrying around the location of the
store, so use that instead.

Test Plan:
build/bin/test_tensorexpr && build/bin/tensorexpr_bench

[ghstack-poisoned]

bertmaher committed:
clang-format on "Avoid using ReduceOp->output_args() in rfactor"


Summary:

I'm trying to get rid of `output_args` for reductions, because they
shouldn't be necessary; it's reducing over its reduction axis, why
does it need to know where its output is going?

Rfactor is probably the trickiest place where we use output_args, but
it looks like it's mostly just carrying around the location of the
store, so use that instead.

Test Plan:
build/bin/test_tensorexpr && build/bin/tensorexpr_bench

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -2441,7 +2448,7 @@ void LoopNest::rfactor(
 
   auto second_buf = dynamic_cast<const Buf*>(second_reduce->accumulator());
-  std::vector<const Expr*> second_indices = {second_reduce->output_args()};
+  std::vector<const Expr*> second_indices = {old_outer};
```
Replace `second_indices` with `old_outer` below, as noted by clang-tidy?

navahgar reviewed (approved):
LGTM

Can you fix the clang-tidy issues?

bertmaher committed:
fix clang-tidy on "Avoid using ReduceOp->output_args() in rfactor"


Summary:

I'm trying to get rid of `output_args` for reductions, because they
shouldn't be necessary; it's reducing over its reduction axis, why
does it need to know where its output is going?

Rfactor is probably the trickiest place where we use output_args, but
it looks like it's mostly just carrying around the location of the
store, so use that instead.

Test Plan:
build/bin/test_tensorexpr && build/bin/tensorexpr_bench

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D26420548](https://our.internmc.facebook.com/intern/diff/D26420548)

[ghstack-poisoned]
